File: keywalk.c
/*
 * This is in no way a finalized product but rather just a tool for evaluation.
 * However, the basic structure here is to fulfill that need and provide a baseline
 * for a real implmentation.
 */
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/stat.h>

int debug;

struct position {
	char x;
	char y;
} __attribute__((packed));

char keys[4][14] = {
	"`1234567890-=\0",
	"\0qwertyuiop[]\\", 
	"\0asdfghjkl;'\0\0",
	"\0zxcvbnm,./\0\0\0"
};
char keys_caps[4][14] = {
	"~!@#$%^&*()_+\0",
	"\0QWERTYUIOP{}|",
	"\0ASDFGHJKL:\"\0\0",
	"\0ZXCVBNM<>?\0\0\0"
};

void print_table()
{
	int x, y, c;
	for (y = 0; y < 4; y++) {
		for (x = 0; x < 14; x++) {
			c = keys[y][x];
			printf(" %c ",  c == '\0' ? ' ' : c);
		}
		printf("\n");
	}
	for (y = 0; y < 4; y++) {
		for (x = 0; x < 14; x++) {
			c = keys_caps[y][x];
			printf(" %c ",  c == '\0' ? ' ' : c);
		}
		printf("\n");
	}
}

int get_position(struct position *p, char c)
{
	for (p->y = 0; p->y < 4; p->y++) {
		for (p->x = 0; p->x < 14; p->x++) {
			if (c == keys[p->y][p->x] || c == keys_caps[p->y][p->x])
				return 1;
		}
	}
	return 0;
}

int check_word(char *word, char *buffer, size_t maxlen)
{
	struct position pos, last; 
	char *ptr;
	int i, x, y, valid, total, distance;
	size_t len;

	total = 0;
	distance = 0;
	valid = 0;
	memset(buffer, 0, maxlen);
	ptr = buffer;
	len = strlen(word);

	memset(&pos, 0, sizeof(pos));
	for (i = 0; i < len && valid < maxlen; i++) {
		/* For now if we ignore unknown characters */
		if (!get_position(&pos, word[i])) {
			if (debug)
				fprintf(stderr, "Invalid character (%x) for password: %s at %d\n", word[i], word, i);
			continue;
		}

		if (valid > 0) {
			x = abs(pos.x - last.x);
			y = abs(pos.y - last.y);
			distance = x + y;
			total += distance;
			*ptr++ = distance;
		}
		last = pos;
		valid++;
	}

	/* potential overflow */
	if (valid == maxlen)
		return -1;

	return ptr - buffer;
}

int count_values(char *distances, int dlen, int value)
{
	int i;
	int count = 0;

	for (i = 0; i < dlen; i++) {
		if (distances[i] == value)
			count++;
	}

	return count;
}

int find_sequences(char *distances, int dlen)
{
	int i, j, k, l;
	int match = 0;
	int seqlen, half;
	int sstart, mstart;
	int max = 1;
	int total = 0;

	/* Search for sequences from 2 to half the size */
	half = dlen / 2;
	for (seqlen = 2; seqlen <= half; seqlen++) {
		/* Start a sliding window starting at the start and working towards the end */
		for (sstart = 0; (sstart + seqlen) < (dlen - seqlen); sstart++) {
			match = 1;
			/* sliding window for the match */
			for (mstart = (sstart + seqlen); mstart < dlen; ) {
				/* match as much as we can */
				for (i = 0; distances[sstart + i] == distances[mstart + i] && i < seqlen && mstart + i < dlen; i++);
				
				/* if we reached seqlen it's a match */
				if (i == seqlen) {
					if (debug) {
					printf("(%d-%d) ", sstart, + sstart + (i-1));
						for (j = sstart; j < sstart + seqlen; j++) 
							printf("%d ", distances[j]);
						printf("  match (%d-%d)\n", mstart, mstart + (i-1));
					}
					match += 1;
					mstart += i;
				} else {
					mstart += i + 1;
				}
			}
			total = seqlen * match;
			if (total > max) {
				if (debug)
					printf("new max: total=%d max=%d seqlen=%d match=%d\n", total, max, seqlen, match);
				max = total;
			}
		}
	}


	return max;
}

void print_summary(char *word, char *distances, int dist_len, int seq_total, int zeros, int ones)
{
	char buffer[16384], *ptr;
	int i, blen, slen;
	float zeros_ratio, ones_ratio, seq_ratio;

	ptr = buffer;
	for (i = 0; i < dist_len; i++) {
		if (i)
			blen = sprintf(ptr, ",%d", distances[i]);
		else
			blen = sprintf(ptr, "%d", distances[i]);
		ptr += blen;
	}

	slen = strlen(word);
	zeros_ratio = ((float) zeros / (float) dist_len);
	ones_ratio = ((float) ones / (float) dist_len);
	seq_ratio = ((float) seq_total / (float) dist_len);
	printf("%s|%s|%d|%d|%.2f|%.2f|%.2f\n", word, buffer, slen, dist_len, zeros_ratio, ones_ratio, seq_ratio);
}

int main(int argc, char **argv)
{
	int i, x, y, total, distance;
	int fd, rc, slen, lseq;
	struct position pos, last;
	size_t len, remaining;
	char *buffer, *ptr, *word;
	char summary[4096];
	struct stat info;
	int good = 0, bad = 0;
	int ones, zeros;
	int c;

	while ((c = getopt(argc, argv, "d")) != -1) {
		switch (c) {
		case 'd':
			debug = 1;
			break;
		default:
			fprintf(stderr, "Invalid optioni: %c\n", c);
			exit(EXIT_FAILURE);
		}
	}

	if (argc - optind != 1)
		return EXIT_FAILURE;

	fd = open(argv[optind], O_RDONLY);
	if (fd == -1)
		return EXIT_FAILURE;

	rc = fstat(fd, &info);
	buffer = malloc(info.st_size);
	if (!buffer)
		return EXIT_FAILURE;
	memset(buffer, 0, info.st_size);

	remaining = info.st_size;
	ptr = buffer;
	while (remaining > 0) {
		len = read(fd, ptr, remaining);
		if (len == -1)
			return errno;
		ptr += len;
		remaining -= len;
	}
	close(fd);

	printf("word|distances|word-len|dist-len|zeros-ratio|ones-ratio|seq-ratio\n");
	ptr = buffer;
	while (*ptr) {
		word = strsep(&ptr, "\n");
		len = strlen(word);
		word[len] = '\0';
		if ((index(word, '|'))) {
			bad++;
			continue;
		}

		slen = check_word(word, summary, sizeof(summary));
		if (slen == -1) {
			bad++;
			continue;
		}

		good++;

		lseq = find_sequences(summary, slen);
		zeros = count_values(summary, slen, 0);
		ones = count_values(summary, slen, 1);
		print_summary(word, summary, slen, lseq, zeros, ones);
	}
}

File: keywalk.py
#!/usr/bin/python

# Initial python code used to generate data for the graphs but is far too slow for large wordlists

import sys
import math

data = [
    "`1234567890-= ",
    " qwertyuiop[]\\",
    " asdfghjkl;   ",
    " zxcvbnm,./   ",
]

data_caps = [
    "~!@#$%^&*()_+ ",
    " QWERTYUIOP{}|",
    ' ASDFGHJKL:"  ',
    " ZXCVBNM<>?   ",
]

def get_position(c):
    global data, data_caps

    for i in range(len(data)):
        for j in range(len(data[i])):
            if c == data[i][j] or c == data_caps[i][j]:
                return i, j
    return -1, -1

def find_sequences_old(s, word):
    d = {}

    for l in range(2, len(word) // 2):
        for i in range(len(s)):
            sub = word[i:i+l]
            d[sub] = word.count(sub)
        if d.get(sub, 0) == 0:
            break
    return d


def find_longest_sequence(s, word):
    d = {}

    # find substrings 2 to half the size
    for l in range(math.floor(len(s) / 2), 2, -1):
        sub = s[0:l]
        k = 0
        match = True
        for i in range(l, len(s) - l):
            if sub[k] != s[i]:
                match = False
                break
        if match:
            return l

    return 1

def find_sequences(s, word):
    d = {}

    substr = ''
    for x in s:
        substr += '%d,' % (x)

    # find substrings 2 to half the size
    #for l in range(2, math.ceil(len(s) / 2) + 1):
    for l in range(math.floor(len(s) / 2), 1, -1):
        for i in range(0, len(s) - l):
            sub = s[i:i+l]
            substr = repr(sub)
            if d.get(substr):
                continue
            j = i + l
            while j < len(s):
                if sub == s[j:j+l]:
                    d[substr] = d.get(substr, 1) + 1
                    #print('%s l=%d j=%d i=%d %d %d' % (substr, l, j, i, j, j+l))
                    j += l
                else:
                    j += 1

    return d

def check_word(word):
    if word.find('|') != -1:
        return
    if len(word) == 0:
        return 
    data = ''
    data = '%s|' % word
    total = 0

    lx = 0
    ly = 0
    s = []
    for i in range(len(word)):
        y, x = get_position(word[i])
        if x == -1 or y == -1:
            continue

        if i > 0:
            xd = abs(x - lx)
            yd = abs(y - ly)
            distance = xd + yd
            total += distance
            if i > 1:
                data += ',%d' % (distance)
            else:
                data += '%d' % (distance)
            s.append(distance)
            
        lx = x
        ly = y

    #data += ') len=%d' % (len(word))
    #data += ' total=%d' % (total)
    if len(s) > 0:
        data += '|%.1f' % (total / len(s))

        ones = s.count(1)
        data += '|%.1f' % (ones / len(s))

        i = 0
        current = 0
        maxrun = 0
        for i in range(len(s)):
            if s[i] <= 1:
                current += 1
                if current > maxrun:
                    maxrun = current
            else:
                current = 0

        data += '|%d' % (maxrun)
        #data += ' mean=%d' % (total / len(s))
    else:
        data += '|NA|NA|NA'

    d = find_sequences(s, word)
    keys = d.keys()
    sorted(keys, key=lambda x: d[x])
    maxscore = 0
    for k in keys:
        z = len(k.split(','))
        #print('%s %d x %d = %d' % (k, z, d[k], z * d[k]))
        score = len(k.split(',')) * d[k]
        if score > maxscore:
            maxscore = score
        # used for unique sequences
        # col = k.split()
        #data += ' seq(%d)=%d' % (k, d[k])
        #score += d[k] * k
        #print('\t', k, d[k])
    #data += ' score=%.2f' % (score / len(word))
    longest = find_longest_sequence(s, word)
    data += '|%d|%d|%d|%.2f' % (longest, len(word), total, maxscore / len(word))
    print(data)


def main():
    words = []

    buffer = ''
    with open(sys.argv[1], 'rb') as f:
        buffer = f.read()
    for line in buffer.decode('utf-8', errors='ignore').split('\n'):
        words.append(line)

    print('password|path|mean|ones|maxrun|longest|len|total|score')
    for word in words:
        check_word(word)


if __name__ == '__main__':
    main()

